generator/json-generator.lua

```diff
@@ -1,135 +1,162 @@
 -- Constants
 -- ------------------------------------------------
 
-local APOSTROPHE =  '"'
-local COMMA      =  ','
-local LINE_BREAK = '\n'
-local TAB        = '  '
-
-local PARAM  = '${%d:%s (%s)}'
-
 local OUTPUT_FILE = 'love-completions.json'
-
 local WIKI_URL = 'https://love2d.org/wiki/'
 
+local LOVE_MODULE_STRING = 'love.'
+
+-- JSON Output control
+local DEBUG = false
+local KEY_ORDER = {
+    'type',
+    'description',
+    'fields',
+    'args',
+    'returnTypes',
+    'link'
+}
+
 -- ------------------------------------------------
 -- Local Functions
 -- ------------------------------------------------
 
-local function generateArguments( arguments, start )
-    local params = ''
-    start = start or 0
-
-    if arguments then
-        for i, args in ipairs( arguments ) do
-            params = params .. string.format( PARAM, start + i, args.name, args.type )
-
-            -- Add a separator unless we are dealing with the last argument of the function.
-            if i ~= #arguments then
-                params = params .. ', '
-            end
+---
+-- Generates a sequence containing subtables of all the arguments
+--
+local function createArguments( args )
+    local arguments = {}
+    for i, v in ipairs( args ) do
+        arguments[i] = {}
+        arguments[i].name = v.name
+
+        -- Use [foo] notation as display name for optional arguments.
+        if v.default then
+            arguments[i].displayName = string.format( '[%s]', v.name )
         end
     end
+    return arguments
+end
 
-    return params
+local function getType( type )
+    if type == 'string'
+    or type == 'number'
+    or type == 'table'
+    or type == 'boolean'
+    or type == 'function' then
+        return { type = type }
+    end
+    return { type = 'ref', name = type }
 end
 
-local function cleanUpString( str )
-    return str:gsub( '\n\n', ' ' ):gsub( '\"', '\\"' ):gsub( '\0', '' )
+local function createReturnTypes( returns )
+    local returnTypes = {}
+    for i, v in ipairs( returns ) do
+        returnTypes[i] = getType( v.type )
+    end
+    return returnTypes
 end
 
-local function buildCallbackCompletion( f )
-    local arguments = generateArguments( f.variants[1].arguments )
-    return {
-        TAB .. '{' .. LINE_BREAK,
-        TAB .. TAB .. '"displayText": "love.' .. f.name .. APOSTROPHE .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"type": "function"' .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"description": ' .. APOSTROPHE .. cleanUpString( f.description ) .. APOSTROPHE .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"descriptionMoreURL": ' .. APOSTROPHE .. WIKI_URL .. 'love.' .. f.name .. APOSTROPHE .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"snippet": ',
-        APOSTROPHE .. 'love.' .. string.format( f.name .. '(%s)', arguments ) .. APOSTROPHE .. LINE_BREAK,
-        TAB .. '},' .. LINE_BREAK
-    }
+-- TODO handle functions with neither args nor return variables.
```

Comment:
Not really. If you type `love.audio.stop` for example, your cursor will be placed between the brackets of the function `love.audio.stop(|)` instead of behind them `love.audio.stop()|`.

I'm not sure if this is something we can change or if this is an "issue" with autocomplete-lua. 

It's not that big of an issue - just a small annoyance.